File: client/bcephal-client-model/Grids/GrilleColumn.cs
using Bcephal.Models.Base;
using Bcephal.Models.Dimensions;
using Bcephal.Models.Filters;
using Bcephal.Models.Utils;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Media;

namespace Bcephal.Models.Grids
{
	[Serializable]
	public class GrilleColumn : Persistent
	{

		public String Name { get; set; }

		public DimensionType Type { get; set; }

		public long? DimensionId { get; set; }

		public string DimensionName { get; set; }

		public string DimensionFunction { get; set; }

		public string DimensionFormat { get; set; }

		[JsonIgnore]
		public MeasureFunctions MeasureFunction
		{
			get { return MeasureFunctions.GetByCode(this.DimensionFunction); }
			set { this.DimensionFunction = value != null ? value.code : null; }
		}


		[JsonIgnore]
		public Grille Grid { get; set; }

		public GrilleColumnCategory Category;

		public bool Mandatory { get; set; }

		public bool Editable { get; set; }

		public bool ShowValuesInDropList { get; set; }

		public int Position { get; set; }

		public bool Show { get; set; }

		public int? BackgroundColor { get; set; }

		public int? ForegroundColor { get; set; }

		public int? Width { get; set; }

		public string FixedType { get; set; }

		[JsonIgnore]
		public GrilleColumnFixedStyle ColumnFixedStyle
		{
			get { return GrilleColumnFixedStyle.GetByCode(this.FixedType); }
			set { this.FixedType = value != null ? value.code : null; }
		}

		public string DefaultStringValue { get; set; }

		public decimal? DefaultDecimalValue { get; set; }

		public DateTime? DefaultDateValue { get; set; }

		public bool ApplyDefaultValueIfCellEmpty { get; set; }

		public bool ApplyDefaultValueToFutureLine { get; set; }

		public DimensionFormat Format { get; set; }

		public bool? OrderAsc { get; set; }

		public string GroupBy { get; set; }

		//[JsonIgnore]
		//public PeriodGrouping PeriodGroupBy
		//{
		//	get { return PeriodGrouping.GetByCode(this.GroupBy); }
		//	set { this.GroupBy = value != null ? value.code : null; }
		//}

		[JsonIgnore]
		public object DefaultValue
		{
			get
			{
				if (this.IsMeasure)
				{
					return this.DefaultDecimalValue;
				}
				if (this.IsPeriod)
				{
					return this.DefaultDateValue;
				}
				return this.DefaultStringValue;
			}
			set
			{
				this.DefaultDateValue = null;
				this.DefaultDecimalValue = null;
				this.DefaultStringValue = null;
				if (this.IsMeasure)
				{
					this.DefaultDecimalValue = (decimal?)value;
				}
				else if (this.IsPeriod)
				{
					this.DefaultDateValue = (DateTime?)value;
				}
				else
				{
					this.DefaultStringValue = (string)value;
				}
			}
		}

		[JsonIgnore]
		public string TypeAsString { get { return this.Type.ToString(); } set { } }

		[JsonIgnore]
		public string CategoryAsString { get { return this.Category.ToString(); } set { } }

		[JsonIgnore]
		public string ColumnName { get { return RangeUtil.GetColumnName(this.Position + 1); } set { } }

   //     [JsonIgnore]
   //     public Color Background
   //     {
   //         get
   //         {
   //             return this.BackgroundColor.HasValue ?
   //             new SolidColorBrush(ColorToLongConverter.DecimalToColor(this.BackgroundColor.Value)).Color
   //                 : GrilleColumnStyle.DEFAULT_BACKGROUND.Color;
   //         }
   //         set 
			//{ 
			//	this.BackgroundColor = ColorToLongConverter.ColorToDecimal(value);
			//}
   //     }

   //     [JsonIgnore]
   //     public Color Foreground
   //     {
   //         get
   //         {
   //             return this.ForegroundColor.HasValue ?
   //                 new SolidColorBrush(ColorToLongConverter.DecimalToColor(this.ForegroundColor.Value)).Color
   //               : GrilleColumnStyle.DEFAULT_FOREGROUND.Color;
   //         }
   //         set 
			//{
			//	this.ForegroundColor = ColorToLongConverter.ColorToDecimal(value);
			//}
   //     }


		[JsonIgnore]
		public string DefaultStringColorB = "#adb9ca";
		[JsonIgnore]
		public string DefaultStringColorF = "#ffffff";

		[JsonIgnore]
		public string Backgrounds
		{
			get
			{
				return BackgroundColor.HasValue ? convertToHex(BackgroundColor.Value) : DefaultStringColorB;
			}
			set
			{
				BackgroundColor = Convert.ToInt32(value.Substring(1), 16);
			}
		}
		[JsonIgnore]
		public string Foregrounds
		{
			get
			{
				return ForegroundColor.HasValue ? convertToHex(ForegroundColor.Value) : DefaultStringColorF;
			}
			set
			{
				ForegroundColor = Convert.ToInt32(value.Substring(1), 16);
			}
		}

		private string convertToHex(int color)
		{
			string hex = color.ToString("X");
			while (hex.Length < 6)
			{
				hex = "0" + hex;
			}
			return "#" + hex;
		}

		[JsonIgnore] public bool IsAttribute { get { return this.Type == DimensionType.ATTRIBUTE; } }

		[JsonIgnore] public bool IsMeasure { get { return this.Type == DimensionType.MEASURE; } }

		[JsonIgnore] public bool IsPeriod { get { return this.Type == DimensionType.PERIOD; } }


		public GrilleColumn()
        {
			this.Type = DimensionType.ATTRIBUTE;
			this.Category = GrilleColumnCategory.USER;
			this.Show = true;
			this.Editable = true;
			this.ShowValuesInDropList = false;
			this.Format = new DimensionFormat();
		}


		public override int CompareTo(object obj)
		{
			if (obj == null || !(obj is GrilleColumn)) return 1;
			if (this == obj) return 0;
			if (this.Id.HasValue && this.Id.Equals(((GrilleColumn)obj).Id)) return 0;
			if (this.Position.Equals(((GrilleColumn)obj).Position))
			{
				return this.Name.CompareTo(((GrilleColumn)obj).Name);
			}
			return this.Position.CompareTo(((GrilleColumn)obj).Position);
		}

		public override string ToString()
		{
			return this.Name;
		}


	}
}

File: server/bcephal-grid/src/main/java/com/moriset/bcephal/grid/domain/GrilleColumn.java
/**
 * 
 */
package com.moriset.bcephal.grid.domain;

import java.math.BigDecimal;
import java.util.Date;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.moriset.bcephal.domain.DataSourceType;
import com.moriset.bcephal.domain.HorizontalAlignment;
import com.moriset.bcephal.domain.Persistent;
import com.moriset.bcephal.domain.dimension.Attribute;
import com.moriset.bcephal.domain.dimension.CalculatedMeasure;
import com.moriset.bcephal.domain.dimension.Dimension;
import com.moriset.bcephal.domain.dimension.DimensionFormat;
import com.moriset.bcephal.domain.dimension.DimensionType;
import com.moriset.bcephal.domain.dimension.Measure;
import com.moriset.bcephal.domain.dimension.Period;
import com.moriset.bcephal.domain.filters.PeriodGrouping;
import com.moriset.bcephal.utils.JsonDateDeserializer;
import com.moriset.bcephal.utils.JsonDateSerializer;

import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Inheritance;
import jakarta.persistence.InheritanceType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * @author B-Cephal Team
 *
 *         29 mars 2021
 */
@Inheritance(strategy = InheritanceType.JOINED)
@Entity(name = "GrilleColumn")
@Table(name = "BCP_GRID_COLUMN")
@Data 
@EqualsAndHashCode(callSuper = false)
public class GrilleColumn extends Persistent {

	private static final long serialVersionUID = -3541296295134746100L;

	public static final String DEFAULT_SQL_COL_VAR_NAME = "item";
		
	@Id
	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "grid_column_seq")
	@SequenceGenerator(name = "grid_column_seq", sequenceName = "grid_column_seq", initialValue = 1,  allocationSize = 1)
	private Long id;	

	private String name;
	
	@Enumerated(EnumType.STRING)
	private DimensionType type;
	
	@Enumerated(EnumType.STRING)
	private HorizontalAlignment alignment;
	
	private Long dimensionId;
	
	private String dimensionName;
	
	private String dimensionFunction;
	
	private String dimensionFormat;
	
	@ToString.Exclude @EqualsAndHashCode.Exclude
	@JsonIgnore
	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "grid")
	private Grille grid;
	
	@Enumerated(EnumType.STRING) 
	private GrilleColumnCategory category;
	
	private boolean mandatory;
	
	private boolean editable;	
	
	private boolean showValuesInDropList;
	
	private int position;

	private boolean show;	
	
	private Integer backgroundColor;

	private Integer foregroundColor;
	
	private Integer width;
	
	private String fixedType;
		
	private String defaultStringValue;
	
	private BigDecimal defaultDecimalValue;
	
	@JsonSerialize(using = JsonDateSerializer.class)
	@JsonDeserialize(using = JsonDateDeserializer.class)
	private Date defaultDateValue;
	
	private boolean applyDefaultValueIfCellEmpty;
	
	private boolean applyDefaultValueToFutureLine;
	
	
	@Embedded	
	private DimensionFormat format;
	
	@Enumerated(EnumType.STRING) 
	private PeriodGrouping groupBy;
	
	@Transient
	private Boolean orderAsc;
	
	@Transient @JsonIgnore
	private DataSourceType dataSourceType;
	
	@Transient @JsonIgnore
	private Long dataSourceId;
	
	@Transient @JsonIgnore
	private String alias;
	
	@Transient @JsonIgnore
	private CalculatedMeasure calculatedMeasure;
	

	public GrilleColumn() {
		this.show = true;	
		this.category = GrilleColumnCategory.USER;
		this.mandatory = false;
		this.format = new DimensionFormat();
		this.editable = true;
		this.dataSourceType = DataSourceType.UNIVERSE;
	}

	public GrilleColumn(Long id, String name, DimensionType type, Long dimensionId, String dimensionName) {
		this();
		this.id = id;
		this.name = name;
		this.type = type;
		this.dimensionId = dimensionId;
		this.dimensionName = dimensionName;
	}
	
	@JsonIgnore
	public boolean isAttribute() {
		return getType() != null && getType().isAttribute();
	}
	
	@JsonIgnore
	public boolean isMeasure() {
		return getType() != null && getType().isMeasure();
	}
	
	@JsonIgnore
	public boolean isCalculatedMeasure() {
		return getType() != null && getType().isCalculatedMeasure();
	}
	
	@JsonIgnore
	public boolean isPeriod() {
		return getType() != null && getType().isPeriod();
	}
	
	@JsonIgnore
	public Object getDefaultValue() {
		if(isAttribute()) {
			return getDefaultStringValue();
		}
		else if(isMeasure()) {
			return getDefaultDecimalValue();
		}
		else if(isPeriod()) {
			return getDefaultDateValue();
		}
		return null;
	}

	@JsonIgnore
	public String getUniverseTableColumnName() {
		if(isAttribute()) {
			return new Attribute(getDimensionId(), getDimensionName(), getDataSourceType(), getDataSourceId()).getUniverseTableColumnName();
		}
		else if(isMeasure()) {
			return new Measure(getDimensionId(), getDimensionName(), getDataSourceType(), getDataSourceId()).getUniverseTableColumnName();
		}
		else if(isPeriod()) {
			return new Period(getDimensionId(), getDimensionName(), getDataSourceType(), getDataSourceId()).getUniverseTableColumnName();
		}
		else if(isCalculatedMeasure() && calculatedMeasure != null) {
			return calculatedMeasure.toSql();
		}
		return null;
	}
	
	@JsonIgnore
	public String getUniverseTableColumnType() {
		if(isAttribute()) {
			return new Attribute().getUniverseTableColumnType();
		}
		else if(isMeasure()) {
			return new Measure().getUniverseTableColumnType();
		}
		else if(isPeriod()) {
			return new Period().getUniverseTableColumnType();
		}
		else if(isCalculatedMeasure()) {
			return new Measure().getUniverseTableColumnType();
		}
		return null;
	}
	
	@JsonIgnore
	public Dimension getDimension() {
		if(isAttribute()) {
			return new Attribute(getDimensionId(), getDimensionName(), getDataSourceType(), getDataSourceId());
		}
		else if(isMeasure()) {
			return new Measure(getDimensionId(), getDimensionName(), getDataSourceType(), getDataSourceId());
		}
		else if(isPeriod()) {
			return new Period(getDimensionId(), getDimensionName(), getDataSourceType(), getDataSourceId());
		}
		return null;
	}
	
	@JsonIgnore
	public String buildDbColumnAlias() {
		return "column_" + getType() + "_" + getDimensionId();
	}
	
	
	@Override
	public String toString() {
		return name != null ? name : super.toString();
	}

	@JsonIgnore
	@Override
	public GrilleColumn copy() {
		GrilleColumn copy = new GrilleColumn();
		copy.setName(this.getName());
		copy.setType(type);
		copy.setDimensionId(dimensionId);
		copy.setDimensionName(dimensionName);
		copy.setDimensionFunction(dimensionFunction);
		copy.setDimensionFormat(dimensionFormat);
		copy.setCategory(category);
		copy.setMandatory(mandatory);
		copy.setEditable(editable);	
		copy.setShowValuesInDropList(showValuesInDropList);
		copy.setPosition(position);
		copy.setShow(show);	
		copy.setBackgroundColor(backgroundColor);
		copy.setForegroundColor(foregroundColor);
		copy.setWidth(width);
		copy.setFixedType(fixedType);
		copy.setDefaultStringValue(defaultStringValue);
		copy.setDefaultDecimalValue(defaultDecimalValue);
		copy.setDefaultDateValue(defaultDateValue);
		copy.setApplyDefaultValueIfCellEmpty(applyDefaultValueIfCellEmpty);
		copy.setApplyDefaultValueToFutureLine(applyDefaultValueToFutureLine);
		copy.setFormat(format != null ? format.copy() : null);
		copy.setGroupBy(getGroupBy());
		return copy;
	}



}
